### PR TITLE
CSS: disconnected elements should be hidden

### PR DESCRIPTION
**Pull request by timmywil**
Fixes gh-3043

### CODE REVIEW

File: test/unit/css.js [hunk 1]
```diff
@@ -1100,6 +1100,9 @@ QUnit.test( ":visible/:hidden selectors", function( assert ) {
 
 	$a = jQuery( "<a href='#'><h1>Header</h1></a>" ).appendTo( "#qunit-fixture" );
 	assert.ok( $a.is( ":visible" ), "Anchor tag with flow content is visible (gh-2227)" );
+
+	assert.ok( !jQuery( "<div>Test</div>" ).is( ":visible" ), "Disconnected element is not visible" );
```
Comment (markelog) [inline test/unit/css.js]: Could those assertions be in separate test?

File: src/css/hiddenVisibleSelectors.js [hunk 2]
```diff
@@ -1,15 +1,21 @@
...
+], function( jQuery, document, support ) {
 
 function getDisplay( elem ) {
 	return elem.style && elem.style.display || jQuery.css( elem, "display" );
 }
 
 function filterHidden( elem ) {
+
+	// Disconnected elements are considered hidden
+	if ( !jQuery.contains( elem.ownerDocument || document, elem ) ) {
```
Comment (markelog) [inline src/css/hiddenVisibleSelectors.js]: Could you remind me when element doesn't have an `ownerDocument` property

Comment (markelog) [inline src/css/hiddenVisibleSelectors.js]: Maybe we could put this in the comment too?

File: test/unit/css.js [hunk 3]
```diff
@@ -1100,6 +1100,9 @@ QUnit.test( ":visible/:hidden selectors", function( assert ) {
...
+	assert.ok( !jQuery( "<div><div>Test</div></div>" ).find("div").is( ":visible" ), "Disconnected element is not visible" );
```
Comment (markelog) [inline test/unit/css.js]: Duplicated name assertion, same as in above test

File: src/css/hiddenVisibleSelectors.js (hunk 2 above)
Comment (timmywil) [inline src/css/hiddenVisibleSelectors.js]: I don't recall all the cases, but this is pretty standard in our codebase. We do it in several places without a comment.

File: test/unit/css.js (hunk 1 above)
Comment (timmywil) [inline test/unit/css.js]: I guess they could, but it seems appropriate in this test.

File: src/css/hiddenVisibleSelectors.js (hunk 2 above)
Comment (timmywil) [inline src/css/hiddenVisibleSelectors.js]: Actually, it may not apply here considering we do a nodeType check in Sizzle.

Comment (timmywil) [inline src/css/hiddenVisibleSelectors.js]: I feel nervous not doing it tho. Seems like whenever I leave it out, I hit a problem eventually.

Comment (markelog) [inline src/css/hiddenVisibleSelectors.js]: I'm usually not in favor of the code which purpose is unclear

**Comment by mgol** [comment]
The patch looks good but we'll most likely need another modification in this code to fix #3065.

**Comment by timmywil** [comment]
Merged in 1.12-stable.

**Pull request closed without merging**